go-controller/pkg/node/controllers/egressip/egressip.go

```diff
@@ -0,0 +1,721 @@
+			return fmt.Errorf("unable to own chain %s: %v", multiNICChain, err)
+		}
+		ruleArgs, err := c.iptablesManager.GetIPv4ChainRuleArgs(utiliptables.TableNAT, multiNICChain)
+		if err != nil {
+			return fmt.Errorf("unable to get chain %s rules: %v", multiNICChain, err)
+		}
+
+		c.removeStaleIPTableRules(ruleArgs, validV4IPTablesRules, utiliptables.ProtocolIPv4)
+
+		if len(validV4IPTablesRules) > 0 {
+			if err = c.iptablesManager.EnsureRules(utiliptables.TableNAT, multiNICChain, utiliptables.ProtocolIPv4, validV4IPTablesRules); err != nil {
+				return fmt.Errorf("failed to ensure rules (%+v) in chain %s: %v", validV4IPTablesRules, multiNICChain, err)
+			}
+			if err = c.iptablesManager.EnsureRules(utiliptables.TableNAT, utiliptables.ChainPostrouting, utiliptables.ProtocolIPv4, jumpRule); err != nil {
+				return fmt.Errorf("failed to create rule in chain %s to jump to chain %s: %v", utiliptables.ChainPostrouting, multiNICChain, err)
+			}
+		}
+	}
+	if c.v6 {
+		if err = c.iptablesManager.OwnChain(utiliptables.TableNAT, multiNICChain, utiliptables.ProtocolIPv6); err != nil {
+			return fmt.Errorf("unable to own chain %s: %v", multiNICChain, err)
+		}
+		ruleArgs, err := c.iptablesManager.GetIPv6ChainRuleArgs(utiliptables.TableNAT, multiNICChain)
+		if err != nil {
+			return fmt.Errorf("unable to get chain %s rules: %v", multiNICChain, err)
+		}
+
+		c.removeStaleIPTableRules(ruleArgs, validV6IPTablesRules, utiliptables.ProtocolIPv6)
+
+		if len(validV6IPTablesRules) > 0 {
+			if err = c.iptablesManager.EnsureRules(utiliptables.TableNAT, multiNICChain, utiliptables.ProtocolIPv6, validV4IPTablesRules); err != nil {
+				return fmt.Errorf("unable to ensure iptables rules: %v", err)
+			}
+			if err = c.iptablesManager.EnsureRules(utiliptables.TableNAT, utiliptables.ChainPostrouting, utiliptables.ProtocolIPv4, jumpRule); err != nil {
+				return fmt.Errorf("unable to ensure iptables rules for jump rule: %v", err)
+			}
+		}
+	}
+	c.removeStaleAddresses(linkAddresses, validEIPAddressess)
+
+	// ensure valid addresses are assigned. Fine to repeat requests.
+	for _, eIP := range validEIPAddressess {
+		if err = c.linkManager.AddAddress(eIP); err != nil {
+			klog.Errorf("Failed to add egress IP %s to link manager: %v", eIP.IP.String(), err)
+		}
+	}
+	// cleanup stale routes in the table IDs we manage
+	c.removeStaleRoutes(existingLinkRoutes, validRoutesPerLinks)
+	// ensure what we expect is present. It's safe to send multiple equal adds
+	for _, validRoutesPerLink := range validRoutesPerLinks {
+		c.routeManager.Add(validRoutesPerLink)
+	}
+
+	c.removeStaleRules(existingRules, validRules)
+	// ensure what we expect is present. It's safe to send multiple adds
+	for _, validRule := range validRules {
+		if err = c.ruleManager.AddRule(validRule); err != nil {
+			klog.Errorf("Failed to add rule %v to rule manager: %v", validRule, err)
+		}
+	}
+	return nil
+}
+
+func (c *Controller) removeStaleRules(existingRules, validRules []netlink.Rule) {
```

Comment:
small suggestion: If you would use sets you call call diff on them without having to loop over all of the elements of the array, similar to: https://github.com/ovn-org/ovn-kubernetes/blob/96b78039b210bfbadcf39de1ed468f9cd0328c99/go-controller/pkg/ovn/controller/egressservice/egressservice_zone.go#L741

Reply:
Agreed - Ill rework to suit this eventually but not in this PR. Thanks Pat.

Reply:
Fixed.